opentreemap/treemap/udf.py

```diff
@@ -966,156 +1193,149 @@ def _base_collection_fields(self, clean):
         udf = self._get_udf_or_error(key)
 
         if udf.iscollection:
-            self.instance.dirty_collection_udfs[key] = True
-            # HStoreDict cleans values in-place, so we need to do a deep-copy
-            self.collection_fields[key] = copy.deepcopy(val)
+            self.instance.dirty_collection_udfs.add(key)
+            self.collection_fields[key] = val
+        elif val == '' or val is None:
+            self.__delitem__(key)
         else:
-            val = udf.reverse_clean(val)
             super(UDFDictionary, self).__setitem__(key, val)
 
+    def __delitem__(self, key):
+        udf = self._get_udf_or_error(key)
 
-class UDFDescriptor(Creator):
-    def __get__(self, obj, type=None):
-        if obj is None:
-            return None
-        # UDFDictionary needs a reference to the model instance to lookup
-        # collection UDFs
-        udf_dict = obj.__dict__[self.field.name]
-        # Workaround for test failure on some dev machines
-        if udf_dict == '':
-            udf_dict = UDFDictionary({}, self.field, obj)
-        udf_dict.instance = obj
-
-        return udf_dict
-
-    def __set__(self, obj, value):
-        value = self.field.to_python(value)
-        if isinstance(value, dict):
-            value = UDFDictionary(
-                value=value, field=self.field, instance=obj
-            )
-        # setattr goes into infinite recursion, so fish in `__dict__`
-        obj.__dict__[self.field.name] = value
-
-
-class UDFField(DictionaryField):
-    # Overridden to convert HStoreDict values to UDFDictionary values
-    def get_default(self):
-        hstore_dict = super(UDFField, self).get_default()
-        if isinstance(hstore_dict, HStoreDict):
-            return UDFDictionary(hstore_dict, self)
-        else:
-            return hstore_dict
-
-    # Overridden to convert HStoreDict values to UDFDictionary values
-    def get_prep_value(self, value):
-        if isinstance(value, dict) and not isinstance(value, UDFDictionary):
-            return UDFDictionary(value, self)
+        if udf.iscollection:
+            self.instance.dirty_collection_udfs.add(key)
+            del self.collection_fields[key]
         else:
-            return value
-
-    # Overridden to convert HStoreDict values to UDFDictionary values
-    def contribute_to_class(self, cls, name):
-        super(UDFField, self).contribute_to_class(cls, name)
-        setattr(cls, self.name, UDFDescriptor(self))
-
-
-class _UDFProxy(UDFField):
-    def __init__(self, name, *args, **kwargs):
-        super(_UDFProxy, self).__init__(*args, **kwargs)
-        self.column = ('udf', name)
-        self.model = None
-
-    def to_python(self, value):
-        return value
-
-
-class UDFModelBase(ModelBase):
-
-    def __new__(clazz, *args, **kwargs):
-        new = super(UDFModelBase, clazz).__new__(clazz, *args, **kwargs)
-
-        orig = new._meta.get_field_by_name
+            super(UDFDictionary, self).__delitem__(key)
```

Comment:
I have an instance with a choice field and a multichoice field on tree. When I attempt to add a new planting site via the iOS app without filling in any fields the iOS app is `POST`ing a dictionary like this:

```
{
    "plot": {
        "address_street": "158 E 67th St",
        "address_city": "Los Angeles",
        "address_zip": "90003",
        "geom": {
            "y": 33.97840416393667,
            "srid": 4326,
            "x": -118.2718002275569
        }
    },
    "tree": {
        "udf:Tree Choice": "",
        "udf:Tree Multi": []
    }
}
```

The app should probably not be sending empty values like this, but it is.

An exception is raised on this line

```
ERROR:django.request:Internal Server Error: /api/v4/instance/jw-2017-05-23/plots
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/opt/app/core/opentreemap/api/decorators.py", line 107, in wrapperf
    return view_f(request, *args, **kwargs)
  File "/opt/app/core/opentreemap/api/decorators.py", line 151, in wrapper
    return view_f(request, *args, **kwargs)
  File "/opt/app/core/opentreemap/treemap/decorators.py", line 46, in wrapper
    return view_fn(request, instance, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/django_tinsel/decorators.py", line 125, in newreq
    outp = req_function(request, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/django_tinsel/decorators.py", line 106, in newreq
    outp = req_function(request, *args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/django_tinsel/decorators.py", line 33, in routed
    return req_method(request, *args2, **kwargs2)
  File "/opt/app/core/opentreemap/api/decorators.py", line 122, in wrapperf
    return view_f(request, *args, **kwargs)
  File "/opt/app/core/opentreemap/treemap/decorators.py", line 219, in wrapper
    return view_fn(request, instance, *args, **kwargs)
  File "/opt/app/core/opentreemap/api/plots.py", line 30, in wrapper
    plot_dict = plot_update_fn(request, *args, **kwargs)
  File "/opt/app/core/opentreemap/api/plots.py", line 113, in update_or_create_plot
    plot, __ = update_map_feature(data, request.user, plot)
  File "/usr/local/lib/python2.7/dist-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/opt/app/core/opentreemap/treemap/views/map_feature.py", line 318, in update_map_feature
    set_attr_on_model(model, field, value)
  File "/opt/app/core/opentreemap/treemap/views/map_feature.py", line 259, in set_attr_on_model
    model.udfs[udf_name] = val
  File "/opt/app/core/opentreemap/treemap/udf.py", line 1235, in __setitem__
    self.__delitem__(key)
  File "/opt/app/core/opentreemap/treemap/udf.py", line 1246, in __delitem__
    super(UDFDictionary, self).__delitem__(key)
KeyError: u'Tree Choice'
```

This test replicates the problem. I tried it out in `treemap/tests/test_map_feature.py` but I was not sure if that is the best place to put it.

```python
class UpdateMapFeature(OTMTestCase):
    def setUp(self):
        self.instance = make_instance()
        self.user = make_commander_user(self.instance)
        set_write_permissions(self.instance, self.user,
                              'Plot', ['udf:Test choice'])

        UserDefinedFieldDefinition.objects.create(
            instance=self.instance,
            model_type='Plot',
            datatype=json.dumps({'type': 'choice',
                                 'choices': ['a', 'b', 'c']}),
            iscollection=False,
            name='Test choice')

    def test_creates_plot_with_empty_string_choice(self):
        plot = Plot(instance=self.instance)
        request_dict = {'plot.geom': {'x': 0, 'y': 0},
                        'plot.udf:Test choice': ''}
        update_map_feature(request_dict, self.user, plot)
```